File: Management/Services/Canvas/CanvasService.cs
using CanvasModel;
using CanvasModel.Assignments;
using CanvasModel.Courses;
using CanvasModel.EnrollmentTerms;
using CanvasModel.Modules;
using CanvasModel.Pages;
using RestSharp;

namespace Management.Services.Canvas;

public interface ICanvasService
{
  ICanvasAssignmentService Assignments { get; }
  ICanvasAssignmentGroupService AssignmentGroups { get; }
  ICanvasModuleService Modules { get; }
  ICanvasQuizService Quizzes { get; }
  ICanvasCoursePageService Pages { get; }
  Task<IEnumerable<EnrollmentTermModel>> GetTerms();
  Task<IEnumerable<CourseModel>> GetCourses(ulong termId);
  Task<CourseModel> GetCourse(ulong courseId);
  Task<IEnumerable<EnrollmentTermModel>> GetCurrentTermsFor(DateTime? queryDate = null);
  Task UpdateModuleItem(ulong canvasCourseId, ulong canvasModuleId, CanvasModuleItem item);
  Task CreateModuleItem(ulong canvasCourseId, ulong canvasModuleId, string title, string type, ulong contentId);
  Task CreateModuleItem(ulong canvasCourseId, ulong canvasModuleId, string title, string type, string contentId);
  Task CreatePageModuleItem(ulong canvasCourseId, ulong canvasModuleId, string title, CanvasPage canvasPage);
}

public class CanvasService(
  IWebRequestor webRequestor,
  CanvasServiceUtils utils,
  ICanvasAssignmentService Assignments,
  ICanvasAssignmentGroupService AssignmentGroups,
  ICanvasModuleService Modules,
  ICanvasQuizService Quizzes,
  ICanvasCoursePageService Pages,
  MyLogger<ICanvasService> logger
) : ICanvasService
{
  private readonly IWebRequestor webRequestor = webRequestor;
  private readonly CanvasServiceUtils utils = utils;
  private readonly MyLogger<ICanvasService> logger = logger;

  public ICanvasAssignmentService Assignments { get; } = Assignments;
  public ICanvasAssignmentGroupService AssignmentGroups { get; } = AssignmentGroups;
  public ICanvasModuleService Modules { get; } = Modules;
  public ICanvasQuizService Quizzes { get; } = Quizzes;
  public ICanvasCoursePageService Pages { get; } = Pages;

  public async Task<IEnumerable<EnrollmentTermModel>> GetTerms()
  {
    var url = $"accounts/10/terms";

    var request = new RestRequest(url);
    var termResponses = await utils.PaginatedRequest<RedundantEnrollmentTermsResponse>(request);
    var terms = termResponses.Select(r => r.EnrollmentTerms).SelectMany(s => s).ToArray();
    return terms;
  }

  public async Task<IEnumerable<CourseModel>> GetCourses(ulong termId)
  {
    var url = $"courses";
    var request = new RestRequest(url);
    var coursesResponse = await utils.PaginatedRequest<IEnumerable<CourseModel>>(request);
    return coursesResponse.SelectMany(c => c).Where(c => c.EnrollmentTermId == termId).ToArray();
  }

  public async Task<CourseModel> GetCourse(ulong courseId)
  {
    var url = $"courses/{courseId}";
    var request = new RestRequest(url);
    var (data, response) = await webRequestor.GetAsync<CourseModel>(request);

    if (data == null)
    {
      logger.Error(response?.Content ?? "");
      logger.Error(response?.ResponseUri?.ToString() ?? "");
      throw new Exception("error getting course from canvas");
    }
    return data;
  }



  public async Task<IEnumerable<EnrollmentTermModel>> GetCurrentTermsFor(
    DateTime? _queryDate = null
  )
  {
    DateTime queryDate = _queryDate ?? DateTime.Now;

    var terms = await GetTerms();

    var currentTerms = terms
      .Where(t => t.EndAt != null && t.EndAt > queryDate && t.EndAt < queryDate.AddYears(1))
      .Take(3)
      .OrderBy(t => t.StartAt);

    return currentTerms;
  }

  public async Task UpdateModuleItem(
    ulong canvasCourseId,
    ulong canvasModuleId,
    CanvasModuleItem item
  )
  {
    logger.Log($"updating module item {item.Title}");
    var url = $"courses/{canvasCourseId}/modules/{canvasModuleId}/items/{item.Id}";
    var body = new { module_item = new { title = item.Title, position = item.Position } };
    var request = new RestRequest(url);
    request.AddBody(body);

    var (newItem, response) = await webRequestor.PutAsync<CanvasModuleItem>(request);
    if (newItem == null)
      throw new Exception("something went wrong updating module item");
  }

  public async Task CreateModuleItem(
    ulong canvasCourseId,
    ulong canvasModuleId,
    string title,
    string type,
    ulong contentId
  )
  {
    logger.Log($"creating new module item {title}");
    var url = $"courses/{canvasCourseId}/modules/{canvasModuleId}/items";
    var body = new
    {
      module_item = new
      {
        title,
        type = type.ToString(),
        content_id = contentId,
      }
    };
    var request = new RestRequest(url);
    request.AddBody(body);

    var (newItem, _response) = await webRequestor.PostAsync<CanvasModuleItem>(request);
    if (newItem == null)
      throw new Exception("something went wrong updating module item");
  }
  public async Task CreateModuleItem(
    ulong canvasCourseId,
    ulong canvasModuleId,
    string title,
    string type,
    string contentId
  )
  {
    logger.Log($"creating new module item {title}");
    var url = $"courses/{canvasCourseId}/modules/{canvasModuleId}/items";
    var body = new
    {
      module_item = new
      {
        title,
        type = type.ToString(),
        content_id = contentId,
      }
    };
    var request = new RestRequest(url);
    request.AddBody(body);

    var (newItem, _response) = await webRequestor.PostAsync<CanvasModuleItem>(request);
    if (newItem == null)
      throw new Exception("something went wrong updating module item with string content id");
  }
  public async Task CreatePageModuleItem(
    ulong canvasCourseId,
    ulong canvasModuleId,
    string title,
    CanvasPage canvasPage
  )
  {
    logger.Log($"creating new module item {title}");
    var url = $"courses/{canvasCourseId}/modules/{canvasModuleId}/items";
    var body = new
    {
      module_item = new
      {
        title,
        type = "Page",
        page_url = canvasPage.Url,
      }
    };
    var request = new RestRequest(url);
    request.AddBody(body);

    var (newItem, _response) = await webRequestor.PostAsync<CanvasModuleItem>(request);
    if (newItem == null)
      throw new Exception("something went wrong updating module item with string content id");
  }
}

File: nextjs/src/services/canvas/canvasService.ts
import { CanvasEnrollmentTermModel } from "@/models/canvas/enrollmentTerms/canvasEnrollmentTermModel";
import { canvasServiceUtils } from "./canvasServiceUtils";
import { CanvasCourseModel } from "@/models/canvas/courses/canvasCourseModel";
import { CanvasModuleItem } from "@/models/canvas/modules/canvasModuleItems";
import { CanvasPage } from "@/models/canvas/pages/canvasPageModel";
import { CanvasEnrollmentModel } from "@/models/canvas/enrollments/canvasEnrollmentModel";
import { axiosClient } from "../axiosUtils";

const baseCanvasUrl = "https://snow.instructure.com/api/v1";

const getAllTerms = async () => {
  const url = `${baseCanvasUrl}/accounts/10/terms`;
  const { data } = await axiosClient.get<
    {
      enrollment_terms: CanvasEnrollmentTermModel[];
    }[]
  >(url);
  const terms = data.flatMap((r) => r.enrollment_terms);
  return terms;
};

export const canvasService = {
  getAllTerms,
  async getCourses(termId: number) {
    const url = `${baseCanvasUrl}/courses`;
    const response = await axiosClient.get<CanvasCourseModel[][]>(url);
    const allCourses = response.data;
    const coursesInTerm = allCourses
      .flatMap((l) => l)
      .filter((c) => c.enrollment_term_id === termId);
    return coursesInTerm;
  },

  async getCourse(courseId: number): Promise<CanvasCourseModel> {
    const url = `${baseCanvasUrl}/courses/${courseId}`;
    const { data } = await axiosClient.get<CanvasCourseModel>(url);
    return data;
  },

  async getCurrentTermsFor(queryDate: Date = new Date()) {
    const terms = await getAllTerms();
    const currentTerms = terms
      .filter(
        (t) =>
          t.end_at &&
          new Date(t.end_at) > queryDate &&
          new Date(t.end_at) <
            new Date(queryDate.setFullYear(queryDate.getFullYear() + 1))
      )
      .sort(
        (a, b) =>
          new Date(a.start_at ?? "").getTime() -
          new Date(b.start_at ?? "").getTime()
      )
      .slice(0, 3);

    return currentTerms;
  },

  async updateModuleItem(
    canvasCourseId: number,
    canvasModuleId: number,
    item: CanvasModuleItem
  ) {
    console.log(`Updating module item ${item.title}`);
    const url = `${baseCanvasUrl}/courses/${canvasCourseId}/modules/${canvasModuleId}/items/${item.id}`;
    const body = {
      module_item: { title: item.title, position: item.position },
    };
    const { data } = await axiosClient.put<CanvasModuleItem>(url, body);

    if (!data) throw new Error("Something went wrong updating module item");
  },

  async createModuleItem(
    canvasCourseId: number,
    canvasModuleId: number,
    title: string,
    type: string,
    contentId: number | string
  ): Promise<void> {
    console.log(`Creating new module item ${title}`);
    const url = `${baseCanvasUrl}/courses/${canvasCourseId}/modules/${canvasModuleId}/items`;
    const body = { module_item: { title, type, content_id: contentId } };
    const response = await axiosClient.post(url, body);
  },

  async createPageModuleItem(
    canvasCourseId: number,
    canvasModuleId: number,
    title: string,
    canvasPage: CanvasPage
  ): Promise<void> {
    console.log(`Creating new module item ${title}`);
    const url = `${baseCanvasUrl}/courses/${canvasCourseId}/modules/${canvasModuleId}/items`;
    const body = {
      module_item: { title, type: "Page", page_url: canvasPage.url },
    };
    await axiosClient.post<CanvasModuleItem>(url, body);
  },

  async getEnrolledStudents(canvasCourseId: number) {
    console.log(`Getting students for course ${canvasCourseId}`);
    const url = `${baseCanvasUrl}/courses/${canvasCourseId}/enrollments?enrollment_type=student`;
    const { data } = await axiosClient.get<CanvasEnrollmentModel[]>(url);

    if (!data)
      throw new Error(
        `Something went wrong getting enrollments for ${canvasCourseId}`
      );

    return data;
  },
};
